src/cactus_client/action/der_controls.py

```diff
@@ -0,0 +1,255 @@
+from datetime import datetime
+import logging
+from http import HTTPMethod
+import re
+from typing import Any, Optional, cast
+
+from cactus_test_definitions.csipaus import CSIPAusResource
+
+from cactus_client.action.server import (
+    client_error_request_for_step,
+    resource_to_sep2_xml,
+    submit_and_refetch_resource_for_step,
+)
+from cactus_client.error import CactusClientException
+from cactus_client.model.context import ExecutionContext
+from cactus_client.model.execution import ActionResult, StepExecution
+from envoy_schema.server.schema.sep2.response import Response, ResponseType
+from envoy_schema.server.schema.sep2.end_device import EndDeviceResponse
+from envoy_schema.server.schema.sep2.der import (
+    DERControlResponse,
+)
+
+from cactus_client.model.resource import StoredResource
+from cactus_client.schema.validator import to_hex32
+from cactus_client.time import utc_now
+
+logger = logging.getLogger(__name__)
+
+
+def get_edev_lfdi_for_der_control(
+    step: StepExecution, context: ExecutionContext, der_ctl: StoredResource, der_control_href: Optional[str]
+) -> Optional[str]:
+    """Helper function to reduce duplicate code. Checks for a non None parent end device lfdi given a DER control"""
+    resource_store = context.discovered_resources(step)
+    edev = resource_store.get_ancestor_of(CSIPAusResource.EndDevice, der_ctl)
+    if edev is None:
+        context.warnings.log_step_warning(
+            step,
+            message=f"Could not find EndDevice parent for DERControl with href {der_control_href}",
+        )
+        return None
+
+    edev_lfdi = cast(EndDeviceResponse, edev.resource).lFDI
+    if edev_lfdi is None:
+        context.warnings.log_step_warning(
+            step,
+            message=f"Could not find EndDevice lfdi parent for DERControl {der_control_href}",
+        )
+        return None
+
+    return edev_lfdi
+
+
+def determine_response_status(
+    event_status: int, sent_responses: list[str], der_control: DERControlResponse, current_time: datetime
+) -> int | None:
```

Comment:
If this just returns ResponseType | None - You remove the need for downstream casting to ResponseType